pubsub/topic.go

```diff
@@ -522,19 +528,32 @@ type PublishResult = ipubsub.PublishResult
 // need to be stopped by calling t.Stop(). Once stopped, future calls to Publish
 // will immediately return a PublishResult with an error.
 func (t *Topic) Publish(ctx context.Context, msg *Message) *PublishResult {
+	var span trace.Span
+	opts := getSpanAttributes(t.String(), msg)
+	ctx, span = t.tracer.Start(ctx, t.String()+" send", opts...)
+	_, span2 := t.tracer.Start(ctx, t.String()+" add to batch", opts...)
+	defer span2.End()
 	r := ipubsub.NewPublishResult()
 	if !t.EnableMessageOrdering && msg.OrderingKey != "" {
 		ipubsub.SetPublishResult(r, "", errors.New("Topic.EnableMessageOrdering=false, but an OrderingKey was set in Message. Please remove the OrderingKey or turn on Topic.EnableMessageOrdering"))
 		return r
 	}
 
+	if span.SpanContext().IsValid() {
+		if msg.Attributes == nil {
+			msg.Attributes = make(map[string]string)
+		}
+		otel.GetTextMapPropagator().Inject(ctx, NewPubsubMessageCarrier(msg))
+	}
+
 	// Calculate the size of the encoded proto message by accounting
 	// for the length of an individual PubSubMessage and Data/Attributes field.
 	msgSize := proto.Size(&pb.PubsubMessage{
 		Data:        msg.Data,
 		Attributes:  msg.Attributes,
 		OrderingKey: msg.OrderingKey,
 	})
+	span.SetAttributes(semconv.MessagingMessagePayloadSizeBytesKey.Int(msgSize))
```

Comment:
Is this already set in `getSpanAttributes` on line 532, above?

Reply:
I think I understand: After injecting the span into the message attributes, the size of the message needs to be updated.